Add new filter sample_over_line and refactor plot_over_line

MatthewFlamm opened the pull request:
This PR enables a new filter sample_over_line, which is a simple refactoring of plot_over_line, but enables the user to retrieve the data instead of just plotting it.  plot_over_line calls sample_over_line, and now only does plotting.

banesullivan commented on pyvista/core/filters.py:
```diff
@@ -1599,14 +1599,58 @@ def decimate_boundary(dataset, target_reduction=0.5):
+        values = sampled.get_array(scalars)
+        distance = sampled['Distance']
+
+        return distance, values
```
I think this filter should return the `PolyData` line that is sampled. This would allow you to then sample all of the point data in the input `dataset` along the line. Thus, just make the retur statement here

```py
        return line.sample(dataset)
```

banesullivan replied:
and it would make this filter more consistent with the `sample` filter as this method is a really just a helper for the filter

banesullivan commented on pyvista/core/filters.py:
```diff
@@ -1646,19 +1690,7 @@ def plot_over_line(dataset, pointa, pointb, resolution=None, scalars=None,
-            field, scalars = dataset.active_scalars_info
-        values = sampled.get_array(scalars)
-        distance = sampled['Distance']
+        distance, values = DataSetFilters.sample_over_line(dataset, pointa, pointb, resolution, scalars)
```
Then this would fetch the `PolyData` line and grab the `Distance` array and use the user-specified `scalars` argument to grab the scalars that are plotted.

banesullivan commented on pyvista/core/filters.py:
```diff
@@ -1599,14 +1599,58 @@ def decimate_boundary(dataset, target_reduction=0.5):
         return dataset.extract_geometry().triangulate().decimate(target_reduction)
 
 
+    def sample_over_line(dataset, pointa, pointb, resolution=None, scalars=None):
```
No scalars argument as all point arrays would be sampled on to the line

imsodin commented on pyvista/core/filters.py:
```diff
@@ -1599,14 +1599,58 @@ def decimate_boundary(dataset, target_reduction=0.5):
+        if resolution is None:
+            resolution = int(dataset.n_cells)
+        if not isinstance(resolution, int) or resolution < 0:
+            raise RuntimeError('`resolution` must be a positive integer, not {}'.format(type(resolution)))
```
I'd expect the resolution checking to happen in `pyvista.Line`.

banesullivan replied:
Ah good point! It would be good to go ahead and move that check over to thee `pyvista.Line` function

MatthewFlamm replied:
I did nothing but refactor the code, so this was original, but it is always good to clean things up.  Let me dig deeper, I suppose this is a chance to learn more about pyvista internals.

banesullivan replied:
Yep, this one is on me when I originally wrote this filter. If you could move that check over to the following function, it would be a help!

https://github.com/pyvista/pyvista/blob/86a45088696f157c9c5751f214d4feb49a1091fa/pyvista/utilities/geometric_objects.py#L326

MatthewFlamm replied:
With the current version if I do

```py
pv.Line([0, 0, 0], [0, 0, 1], resolution=10.5)
```

I get a helpful error already from presumably vtk
```txt
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/redacted/pyvista/pyvista/utilities/geometric_objects.py", line 348, in Line
    src.SetResolution(resolution)
TypeError: SetResolution argument 1: integer argument expected, got float
```

Do you still prefer a check inside pyvista with RunTimeError?

MatthewFlamm replied:
I didn't touch this part yet as I'm not sure what you want.  My understanding is that `ValueError` is for arguments of the wrong type, which seems to fit here.

banesullivan replied:
ah that's a good point - then I would just let that chack from vtk be it

banesullivan replied:
i.e. no need to even have a check inside pyvista

MatthewFlamm replied:
ok more subtleties....

```py
pv.Line([0, 0, 0], [0, 0, 1], resolution=0)
pv.Line([0, 0, 0], [0, 0, 1], resolution=-1)
```

All of these lines return a line with two points.  I think we should still check the magnitude to be positive in `Line`, which is in the newest commit I pushed.

MatthewFlamm commented:
I removed the check for integer resolution, but moved the check for positive resolution (this one isn't checked in vtk) to `Line`.  I added to test_line to make sure that revisions in vtk don't break this behavior.  I also added some a test for the new sample_over_line filter.

banesullivan commented:
@MatthewFlamm, I'm happy with this at this point, are you finished?

@imsodin, does this look good to you now?

MatthewFlamm commented:
I'm done, unless there are other comments.

imsodin commented on pyvista/core/filters.py:
```diff
@@ -1646,21 +1675,16 @@ def plot_over_line(dataset, pointa, pointb, resolution=None, scalars=None,
-        line = pyvista.Line(pointa, pointb, resolution=resolution)
-        sampled = line.sample(dataset)
+        # Sample on line
+        sampled = DataSetFilters.sample_over_line(dataset, pointa, pointb, resolution)
```
This must be dataset.sample_...

More generally maybe there should be a static analyser on CI to catch such things.

banesullivan replied:
I've implemented a lot of filters that call other filters this way. I thought it was "safer"

banesullivan replied:
is this bad practice, @imsodin?

MatthewFlamm replied:
This whole class is littered with both types examples.  What is the preferred approach, and I can use that one?

banesullivan replied:
I'm happy with it as is, but yes, you're right that the class has both examples all over the place. We're working to refactor things like this to make them more consistent on a separate branch, so I say just leave it and we'll choose the "proper" method later

imsodin replied:
Ah I haven't been aware this was used in the same file. Maybe it's just simply lack of experience on my part that i have never seen a class instance method be called like this. The only practice I could come up with that might be somewhat relevant is that in python there should be "one, preferably obvious way of doing things" and if I had to choose I'd definitely choose variable.method(... over class.method(variable, ...

MatthewFlamm replied:
I agree with @imsodin .  I found this [reference](https://docs.python.org/3/tutorial/classes.html#inheritance) that 

> There is a simple way to call the base class method directly: just call BaseClassName.methodname(self, arguments).

This is the strategy used here.  I see this being useful if you don't want derived classes to overwrite the sample_over_line method call inside the plot_over_line method, but you do want to be able to overwrite it in general.

I'm not sure this is what you want however.

imsodin replied:
Regardless of whether or not it should change, that's a topic for a separate PR.
